packages/compartment-mapper/src/types/external.d.ts

```diff
@@ -0,0 +1,350 @@
+  moduleTransforms?: ModuleTransforms;
+};
+
+export type SyncModuleTransformsOption = {
+  syncModuleTransforms?: SyncModuleTransforms;
+};
+
+export type PolicyOption = {
+  policy?: any;
+};
+
+type LanguageForExtensionOption = {
+  languageForExtension?: LanguageForExtension;
+};
+
+// ////////////////////////////////////////////////////////////////////////////////
+// Common option groups:
+
+export type SyncOrAsyncArchiveOptions = Partial<{
+  captureSourceLocation: CaptureSourceLocationHook;
+}> &
+  ParserForLanguageOption &
+  CompartmentOption &
+  PolicyOption;
+
+type SyncOrAsyncImportingOptions = SearchSuffixesOption & SourceMapHookOption;
+
+type ImportingOptions = ModulesOption &
+  SyncOrAsyncImportingOptions &
+  ExitModuleImportHookOption;
+
+type SyncImportingOptions = ModulesOption &
+  SyncOrAsyncImportingOptions &
+  ExitModuleImportNowHookOption;
+
+type LinkingOptions = ParserForLanguageOption &
+  CompartmentOption &
+  LanguageForExtensionOption &
+  SyncModuleTransformsOption &
+  ModuleTransformsOption;
+
+// ////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * The result of {@link captureFromMap}.
+ */
+export type CaptureResult = {
+  captureCompartmentMap: CompartmentMapDescriptor;
+  captureSources: Sources;
+  compartmentRenames: Record<string, string>;
+};
+
+/**
+ * The compartment mapper can capture the Sources for all loaded modules
+ * for bundling, archiving, or analysis.
+ */
+export type Sources = Record<string, CompartmentSources>;
+export type CompartmentSources = Record<string, ModuleSource>;
+
+// TODO unionize:
+export type ModuleSource = Partial<{
+  /** module loading error deferred to later stage */
+  deferredError: string;
+  /** package-relative location */
+  location: string;
+  /** fully qualified location */
+  sourceLocation: string;
+  bytes: Uint8Array;
+  /** in lowercase base-16 (hexadecimal) */
+  sha512: string;
+  parser: Language;
+  /** indicates that this is a reference that exits the mapped compartments */
+  exit: string;
+  /** module for the module */
+  record: StaticModuleType;
+}>;
+
+export type SourceMapHook = (
+  sourceMap: string,
+  details: SourceMapHookDetails,
+) => void;
+export type SourceMapHookDetails = {
+  compartment: string;
+  module: string;
+  location: string;
+  sha512: string;
+};
+
+type ModuleTransforms = Record<string, ModuleTransform>;
+
+type SyncModuleTransforms = Record<string, SyncModuleTransform>;
+
+type ModuleTransform = (
+  ...args: ModuleTransformArguments
+) => Promise<ModuleTransformResult>;
+type SyncModuleTransform = (
+  ...args: ModuleTransformArguments
+) => ModuleTransformResult;
+
+type ModuleTransformArguments = [
+  bytes: Uint8Array,
+  specifier: string,
+  moduleLocation: string,
+  packageLocation: string,
+  params: {
+    sourceMap?: string;
+  },
+];
+
+type ModuleTransformResult = {
+  bytes: Uint8Array;
+  parser: Language;
+  sourceMap?: string;
+};
```

Comment:
I think all of these should be exported, since the `moduleTransforms` etc. options are exposed